File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let maxLength = 0;
    let start = 0;
    const chars = new Array(128).fill(-1);
    
    for (let end = 0; end < s.length; end++) {
        const charCode = s.charCodeAt(end);
        start = Math.max(start, chars[charCode] + 1);
        chars[charCode] = end;
        maxLength = Math.max(maxLength, end - start + 1);
    }
    
    return maxLength;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        
        int maxL = 0;
        int start = 0;
        
        for(int end = 0; end < s.length(); end++){
            char c = s.charAt(end);
            if(map.get(c) != null && map.get(c) >= start){
                start = map.get(c) + 1;
            }
            
            map.put(c,end);
            maxL = Math.max(maxL, end-start+1);
        }
        return maxL;
    }
}